networkx/algorithms/shortest_paths/dense.py

```diff
@@ -86,6 +87,82 @@ def floyd_warshall_numpy(G, nodelist=None, weight="weight"):
     return A
 
 
+def floyd_warshall_tree(G, weight="weight"):
+    from collections import defaultdict
+
+    # dictionary-of-dictionaries representation for dist and pred
+    # use some defaultdict magick here
+    # for dist the default is the floating point inf value
+    dist = defaultdict(lambda: defaultdict(lambda: float("inf")))
+    for u in G:
+        dist[u][u] = 0
+    pred = defaultdict(dict)
+    # initialize path distance dictionary to be the adjacency matrix
+    # also set the distance to self to 0 (zero diagonal)
+    undirected = not G.is_directed()
+    for u, v, d in G.edges(data=True):
+        e_weight = d.get(weight, 1.0)
+        dist[u][v] = min(e_weight, dist[u][v])
+        pred[u][v] = u
+        if undirected:
+            dist[v][u] = min(e_weight, dist[v][u])
+            pred[v][u] = v
+
+    for w in G:
+        dist_w = dist[w]  # for speed
+        # out_w will store the adjacency list of the OUT_W tree of the paper,
+        # it is a tree, parent --> list of children
+        out_w = {parent: [] for parent in G}
+
+        pred_w = pred.get(w, {})  # so that new entry is not created just by reading
+        for v in G:
+            if v == w:
+                continue
+            parent = pred_w.get(v, w)  # path w to v in OUT_W
+            out_w[parent].append(v)
+
+        # dfs order dict and skip dict in practical improvements in the paper
+        stack = [w]
+        dfs_dict = {}  # {node: next node in dfs order}
+        skip_dict = {}  # {node: next node after subtree is skipped}
+
+        node = None
+        while stack:
+            next_node = stack.pop()
+
+            if node is not None:
+                dfs_dict[node] = next_node
+
+            if stack:
+                skip_dict[next_node] = stack[-1]
+            stack.extend(out_w[next_node])
+
+            node = next_node
+
+        # main inner loop starts here
+        for u in G:
+            if u == w:  # small optimization
+                continue
+            dist_u = dist[u]
+            dist_uw = dist_u[w]
+            if dist_uw == float("inf"):  # small optimization
+                continue
+
+            # skip w, as realxation will always fail, and skipped to last
+            v = dfs_dict[w]
+            while v is not None:
+                dist_uwv = dist_uw + dist_w[v]
+                if dist_u[v] > dist_uwv:
+                    dist_u[v] = dist_uwv
+                    # update/new entries to be created in lhs
```

Comment:
This `lhs` confuses me each time i read it.  I think it refers to the lhs of the next line. So maybe "to be created in pred[u]".  anyway, check this comment and maybe replace term "lhs".